content/advent-2019/control-packetflow-tcp-nodelay.md

```diff
@@ -0,0 +1,226 @@
+
+A cross-platform TCP socket option that comes helpful here is `TCP_NODELAY`. When enabled, it practically disables Nagle's algorithm. Instead of coalescing small packets, it sends them to the pipe the soonest as possible.
+In general, Nagle's algorithm's goal is to reduce the number of packets sent to save bandwidth and increase throughput with the trade-off sometimes to introduce increased latency to services. On the other hand, `TCP_NODELAY` might decrease throughput for small writes, but there are ways to [mitigate](https://tinyurl.com/h56qu6y) this by using buffers on the application side.
+
+In Go, `TCP_NODELAY` is enabled by default, but the standard library offers the ability to disable the behavior via the [net.SetNoDelay](https://golang.org/pkg/net/#TCPConn.SetNoDelay) method.
+
+## A small experiment
+
+To observe what's happening at the packet-level, and see the differences in packet arrival, we will use a tiny TCP client/server written in Go. Usually, we have inter-connected services across different regions, but for the sake of the experiment, we will experiment in our local machine. The full source code is also available on [Github](https://github.com/dastergon/gopheracademy-advent2019-tcp-no-delay).
+
+The server code (server.go):
+
+```go
+package main
+
+import (
+    "bufio"
+    "fmt"
+    "log"
+    "net"
+    "strings"
+)
+
+func main() {
+    port := ":" + "8000"
+
+    // Create a listening socket.
+    l, err := net.Listen("tcp", port)
+    if err != nil {
+        log.Fatal(err)
+    }
+    defer l.Close()
+
+    for {
+        // Accept new connections.
+        c, err := l.Accept()
+        if err != nil {
+            log.Println(err)
+            return
+        }
+
+        // Process newly accepted connection.
+        go handleConnection(c)
+    }
+}
+func handleConnection(c net.Conn) {
+    fmt.Printf("Serving %s\n", c.RemoteAddr().String())
+
+    for {
+        // Read what has been sent from the client.
+        netData, err := bufio.NewReader(c).ReadString('\n')
+        if err != nil {
+            log.Println(err)
+            return
+        }
+
+        cdata := strings.TrimSpace(netData)
+        if cdata == "GOPHER" {
+            c.Write([]byte("GopherAcademy Advent 2019!"))
+        }
+
+        if cdata == "EXIT" {
+            break
+        }
+    }
+    c.Close()
+}
+```
+
+The client code (client.go):
+
+```go
+package main
+
+import (
+    "fmt"
+    "log"
+    "net"
+)
+
+func main() {
+    target := "localhost:8000"
+
+    raddr, err := net.ResolveTCPAddr("tcp", target)
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    // Establish a connection with the server.
+    conn, err := net.DialTCP("tcp", nil, raddr)
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    fmt.Println("Sending Gophers down the pipe...")
+
+    for i := 1; i <= 5; i++ {
```

Comment:
Not that it matters, but why "1" and "<= 5" instead of "0" and "< 5", which is usually idiomatic for for loops in Go?

Reply:
good catch!